architecture: add `ci-operator` internals page

bbguimaraes opened the pull request:
I find myself having to spend time swapping this information back into my brain
no matter how many times I've worked with it, this would be very useful in a
similar situation in the future.

Normally, I would be wary of separating code from internal documentation, but I
think a centralized source like this is superior to having to piece together
information from `pkg/defaults` and all the files in `pkg/steps`, including all
the code that is irrelevant most of the time.  I tried to include as few
implementation details as possible so that it should be virtually immutable.

If we feel this is a useful addition and worth the investment, I'll
finish the remaining items in the "TODO" section (I already have a draft
detailing all of them). Otherwise, I'll do it anyway and keep it to
myself =)

[preview](https://deploy-preview-156--zealous-shirley-0f56b3.netlify.app/docs/architecture/ci-operator-steps)

stevekuznetsov reviewed (commented):
Looks like a great doc - perhaps we can add a section at the top to explain how to go from the code to an entry here - where in `defaults.go` for instance the translation between YAML and step happens, how the `Provides()` and `Requires()` are used to bound input and output - just a quick breakdown for someone who's looking to take this document to the next step and start poking at code.

bbguimaraes commented:
Update:

- added links to source code
- added a short source code guide to the beginning of the "step graph" section
- added cross-references in the descriptions
- separated steps into thematic sections

All steps are now documented except for the ones in the "Operator Steps" section and their dependencies in other steps, which I'll work on next.

This is ready for review. I'd appreciate specially if someone who knows more about releases and release payloads (@stevekuznetsov , @petr-muller?) could verify the information in the ["Release Steps"](https://deploy-preview-156--zealous-shirley-0f56b3.netlify.app/docs/architecture/ci-operator-steps/#release-steps) section.

bbguimaraes commented:
Update: all steps are now documented.

I would *specially* appreciate a careful review of the new ["Operator Steps"](https://deploy-preview-156--zealous-shirley-0f56b3.netlify.app/docs/architecture/ci-operator-steps/#operator-steps) section.

stevekuznetsov commented on content/en/docs/architecture/ci-operator-steps.md:
```diff
@@ -0,0 +1,439 @@
+- Inputs: all steps that import or build images
+- Outputs: the signal that all image imports and builds have finished
+
+`ImagesReadyStep` is a synthetic step added to the graph to generate the correct
```
The history is that this was added before `dependencies` and in general solves the problem of "some of the images we build are going to go into the release payload we create, but we don't know which and never cared to find out"

stevekuznetsov reviewed (approved):
/hold

@petr-muller @alvaroaleman feel free to unhold. This looks good

petr-muller commented:
This is great. Thanks for taking the time to write this.
/hold cancel
